### PR TITLE
Set instance_map configuration

### PR DESCRIPTION
**Pull request by amandahla**
### Overview

This PR adds instance_map configuration required by horizontal scaling Synapse architecture.

### Rationale

The [instance_map](https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#instance_map) configuration maps main and workers defined.

### Juju Events Changes

N/A

### Module Changes

State was changed to store instance_map configuration.

### Library Changes

N/A

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Charmhub doc will be updated when horizontal scaling is completely in place.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -122,21 +125,38 @@ def is_main(self) -> bool:
...
 
         Returns:
-            List of peer IP addresses.
+            Instance map configuration as a dict.
         """
-        addresses = []
+        unit_name = self.unit.name.replace("/", "-")
+        app_name = self.app.name
+        model_name = self.model.name
+        addresses = [f"{unit_name}.{app_name}-endpoints.{model_name}.svc.cluster.local"]
```
Comment (javierdelapuente) [inline src/charm.py:137]: Maybe just `addresses = [f"{unit_name}.{app_name}-endpoints"]`? (no need for the rest)

File: src/charm.py [hunk 2]
```diff
@@ -122,21 +125,38 @@ def is_main(self) -> bool:
...
-        logger.debug("peer_addresses values are: %s", str(addresses))
-        return addresses
+                # <unit-name>.<app-name>-endpoints.<model-name>.svc.cluster.local
+                unit_name = u.name.replace("/", "-")
+                address = f"{unit_name}.{app_name}-endpoints.{model_name}.svc.cluster.local"
+                addresses.append(address)
+        logger.debug("addresses values are: %s", str(addresses))
+        instance_map = {}
+        main_unit_name = self.unit.name.replace("/", "-")
+        if self.get_main_unit() is not None and isinstance(self.get_main_unit, str):
```
Comment (javierdelapuente) [inline src/charm.py:149]: `isinstance(self.get_main_unit, str)`? get_main_unit is a function

File: src/charm.py [hunk 3]
```diff
@@ -122,21 +125,38 @@ def is_main(self) -> bool:
...
+            main_unit_name = self.get_main_unit().replace("/", "-")
+        main_unit_address = f"{main_unit_name}.{app_name}-endpoints.{model_name}.svc.cluster.local"
+        for address in addresses:
+            match = re.search(r"-(\d+)", address)
+            if match is not None:
+                unit_number = match.group(1)
+            instance_name = "main" if address == main_unit_address else f"worker{unit_number}"
```
Comment (javierdelapuente) [inline src/charm.py:156]: Two lines before, there was a `if match is not None:`. If it was `None`, then `unit_number` is not defined. Can this happen?

File: src/charm.py [hunk 4]
```diff
@@ -106,13 +107,15 @@ def build_charm_state(self) -> CharmState:
...
-        return CharmState.from_charm(
+        charm_state = CharmState.from_charm(
             charm=self,
             datasource=self._database.get_relation_as_datasource(),
             saml_config=self._saml.get_relation_as_saml_conf(),
             smtp_config=self._smtp.get_relation_as_smtp_conf(),
             redis_config=self._redis.get_relation_as_redis_conf(),
+            instance_map_config=self.instance_map(),
         )
+        return charm_state
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:118]: No particular reason. I must had left while trying a different approach before deciding setting the instance_map in the charm_state ;) I'll change it back.

File: src/synapse/workload.py [hunk 5]
```diff
@@ -435,6 +435,25 @@ def enable_serve_server_wellknown(container: ops.Container) -> None:
...
+    Raises:
+        WorkloadError: something went wrong enabling configuration.
+    """
+    try:
+        config = container.pull(SYNAPSE_CONFIG_PATH).read()
+        current_yaml = yaml.safe_load(config)
+        current_yaml["instance_map"] = charm_state.instance_map_config
+        container.push(SYNAPSE_CONFIG_PATH, yaml.safe_dump(current_yaml))
+    except ops.pebble.PathError as exc:
+        raise WorkloadError(str(exc)) from exc
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/synapse/workload.py:454]: Nice, thanks :)

File: src/charm.py (hunk 3 above)
Comment (amandahla) [reply src/charm.py:158]: The check for match is none is more related to a lint issue than to the possibility of not matching the pattern. At least, this is what is described in the docs. 

"A unit is always named on the pattern <application>/<unit ID>, where <application> is the name of the application and the <unit ID> is its ID number or, for the leader unit, the keyword leader. For example, mysql/0 or mysql/leader. Note: the number designation is a static reference to a unique entity whereas the leader designation is a dynamic reference to whichever unit happens to be elected by Juju to be the leader ."

But I see that this check is confusing so I removed and added the comment with the docs. Thanks!

**Review by javierdelapuente — APPROVED** [review]

**Pull request merged**